File: lab/starter_code/vibe_coding_demo.py
# =============================================================================
# PM599 Week 1 Lab — Vibe Coding Demo (Python version)
# Session: May 26, 2026
# =============================================================================
#
# This script mirrors the R demo but uses Python (pandas, scipy, matplotlib).
# R is the primary language for this course; this file is a secondary reference
# for students more comfortable with Python.
#
# HOW TO USE: Run section by section. At each [AI PROMPT], ask an AI tool
# to help you extend or modify the code as instructed.
# =============================================================================

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scipy.stats as stats

# Set random seed for reproducibility
np.random.seed(42)

# -----------------------------------------------------------------------------
# SECTION 2: Simulate a Dataset
# -----------------------------------------------------------------------------

n = 100  # total sample size
group = ["Treatment"] * (n // 2) + ["Control"] * (n // 2)

sim_data = pd.DataFrame({
    "patient_id":   range(1, n + 1),
    "group":        group,
    "age":          np.round(np.random.normal(55, 10, n)).astype(int),
    "baseline_sbp": np.round(np.random.normal(145, 15, n)).astype(int),
    "followup_sbp": np.concatenate([
        np.round(np.random.normal(130, 12, n // 2)),   # Treatment
        np.round(np.random.normal(140, 12, n // 2))    # Control
    ]).astype(int),
})

sim_data["sbp_change"] = sim_data["followup_sbp"] - sim_data["baseline_sbp"]

print(sim_data.head())
print(sim_data.describe())

# [AI PROMPT 1] ---------------------------------------------------------------
# "Using the sim_data DataFrame, write Python code with pandas to compute the
#  mean and standard deviation of sbp_change separately for Treatment and
#  Control groups."
# Paste AI code here:


# -----------------------------------------------------------------------------
# SECTION 3: Visualize the Outcome
# -----------------------------------------------------------------------------

fig, ax = plt.subplots(figsize=(7, 5))

for i, (grp, color) in enumerate(zip(["Treatment", "Control"], ["#990000", "#FFC72C"])):
    subset = sim_data[sim_data["group"] == grp]["sbp_change"]
    ax.boxplot(subset, positions=[i], widths=0.5, patch_artist=True,
               boxprops=dict(facecolor=color, alpha=0.7))
    ax.scatter(np.random.normal(i, 0.05, len(subset)), subset,
               color=color, alpha=0.4, s=20, zorder=3)

ax.axhline(0, linestyle="--", color="gray", linewidth=1, label="No change")
ax.set_xticks([0, 1])
ax.set_xticklabels(["Treatment", "Control"])
ax.set_ylabel("SBP Change from Baseline (mmHg)")
ax.set_title("Change in Systolic Blood Pressure by Group\n(Simulated data, n=100)")
plt.tight_layout()
plt.show()

# [AI PROMPT 2] ---------------------------------------------------------------
# "Modify the plot above to add the group mean as a red diamond marker on
#  each box, and add a legend."
# Paste AI code here:


# -----------------------------------------------------------------------------
# SECTION 4: Statistical Test
# -----------------------------------------------------------------------------

treatment = sim_data[sim_data["group"] == "Treatment"]["sbp_change"]
control   = sim_data[sim_data["group"] == "Control"]["sbp_change"]

t_stat, p_value = stats.ttest_ind(treatment, control)
print(f"t = {t_stat:.3f}, p = {p_value:.4f}")

# [AI PROMPT 3] ---------------------------------------------------------------
# "Explain the two-sample t-test result above in plain language for a methods
#  section. Include the test statistic, p-value, and interpretation."
# AI explanation (paste as comment):
#


# =============================================================================
# END OF PYTHON DEMO
# Primary course language is R — refer to vibe_coding_demo.R for the full
# version with additional exercises.
# =============================================================================

File: lab/starter_code/vibe_coding_demo.R
# =============================================================================
# PM599 Week 1 Lab — Vibe Coding Demo
# Session: May 26, 2026
# =============================================================================
#
# WHAT IS VIBE CODING?
# Vibe coding means using an AI tool (ChatGPT, Claude, Gemini) to help you
# write, explain, and debug code iteratively. You describe what you want in
# plain language; the AI drafts code; you review, test, and refine together.
#
# THE GOLDEN RULE: You are responsible for every line.
# AI makes mistakes. Always read, understand, and test the code it produces
# before trusting the output.
#
# HOW TO USE THIS SCRIPT:
# Run it section by section. At each [AI PROMPT] comment, try asking an AI
# tool to help you modify or extend the code as instructed.
# =============================================================================

# -----------------------------------------------------------------------------
# SECTION 1: Setup
# -----------------------------------------------------------------------------

# Install packages if needed (run once, then comment out)
# install.packages(c("ggplot2", "dplyr", "survival", "survminer"))

library(ggplot2)
library(dplyr)

# -----------------------------------------------------------------------------
# SECTION 2: Simulate a Dataset
# -----------------------------------------------------------------------------
# We will simulate a simple clinical trial dataset.
# Two treatment groups; continuous outcome (systolic blood pressure at 6 months).

set.seed(42)
n <- 100  # total sample size

sim_data <- data.frame(
  patient_id  = 1:n,
  group       = rep(c("Treatment", "Control"), each = n / 2),
  age         = round(rnorm(n, mean = 55, sd = 10)),
  baseline_sbp = round(rnorm(n, mean = 145, sd = 15)),   # mmHg
  followup_sbp = c(
    round(rnorm(n / 2, mean = 130, sd = 12)),  # Treatment: larger reduction
    round(rnorm(n / 2, mean = 140, sd = 12))   # Control:   smaller reduction
  )
)

# Compute change from baseline
sim_data <- sim_data |>
  mutate(sbp_change = followup_sbp - baseline_sbp)

# Quick look at the data
head(sim_data)
summary(sim_data)

# [AI PROMPT 1] ---------------------------------------------------------------
# "Looking at the sim_data data frame, write R code to compute the mean and
#  standard deviation of sbp_change separately for the Treatment and Control
#  groups. Show the result in a tidy table."
# Paste the AI's code below and test it:

# < paste AI code here >


# -----------------------------------------------------------------------------
# SECTION 3: Visualize the Outcome
# -----------------------------------------------------------------------------

# Box plot of SBP change by group
ggplot(sim_data, aes(x = group, y = sbp_change, fill = group)) +
  geom_boxplot(alpha = 0.7) +
  geom_jitter(width = 0.15, alpha = 0.4, size = 1.5) +
  scale_fill_manual(values = c("Treatment" = "#990000", "Control" = "#FFC72C")) +
  labs(
    title    = "Change in Systolic Blood Pressure by Group",
    subtitle = "Simulated clinical trial data (n = 100)",
    x        = "Group",
    y        = "SBP Change from Baseline (mmHg)",
    fill     = "Group"
  ) +
  theme_minimal(base_size = 13) +
  theme(legend.position = "none")

# [AI PROMPT 2] ---------------------------------------------------------------
# "Modify the ggplot code above to also add a horizontal dashed line at y = 0
#  to mark no change, and add the mean of each group as a red diamond point."
# Paste the AI's code below and test it:

# < paste AI code here >


# -----------------------------------------------------------------------------
# SECTION 4: Statistical Test
# -----------------------------------------------------------------------------

# Two-sample t-test: is the mean SBP change different between groups?
t_result <- t.test(sbp_change ~ group, data = sim_data)
print(t_result)

# [AI PROMPT 3] ---------------------------------------------------------------
# "Explain the output of the t.test() result above in plain language, suitable
#  for a methods section of a clinical paper. Include the test statistic,
#  degrees of freedom, p-value, and 95% confidence interval."
# Paste the AI's explanation below as a comment:

# AI explanation:
#


# -----------------------------------------------------------------------------
# SECTION 5: Simple Linear Regression
# -----------------------------------------------------------------------------

# Does age predict the magnitude of SBP reduction?
lm_fit <- lm(sbp_change ~ age + group, data = sim_data)
summary(lm_fit)

# [AI PROMPT 4] ---------------------------------------------------------------
# "Write R code to create a scatter plot of age (x-axis) vs. sbp_change
#  (y-axis), with separate regression lines for Treatment and Control groups,
#  using ggplot2. Color the points and lines by group using USC colors
#  (#990000 for Treatment, #FFC72C for Control)."
# Paste the AI's code below and test it:

# < paste AI code here >


# -----------------------------------------------------------------------------
# SECTION 6: Your Turn — Modify the Scenario
# -----------------------------------------------------------------------------

# [EXERCISE] ------------------------------------------------------------------
# Use an AI tool to help you modify this analysis for a different scenario.
# Pick ONE of the following and ask the AI to rewrite the simulation:
#
#   a) Change the outcome to a binary variable (e.g., "responded to treatment:
#      yes/no") and run a chi-squared test instead of a t-test.
#
#   b) Add a third group (e.g., "Low Dose", "High Dose", "Control") and use
#      a one-way ANOVA instead of a t-test.
#
#   c) Add a covariate (e.g., sex: male/female) and extend the regression model
#      to include it as an additional predictor.
#
# Suggested prompt template:
# "I have this R simulation code [paste Section 2]. Modify it so that [describe
#  your chosen scenario]. Then update the statistical test accordingly."
#
# Paste your modified code below:

# < paste modified code here >


# =============================================================================
# END OF SCRIPT
# Remember: read, understand, and test every line the AI produces.
# If the code runs but gives unexpected results, that is just as important
# to investigate as code that throws an error.
# =============================================================================
